api/tests/provider_api_tests.py

```diff
@@ -0,0 +1,45 @@
+from django.urls import reverse
+from rest_framework import status
+from rest_framework.test import APITestCase
+from customers.models import Provider
+from customers.tests.factories import ProviderFactory
+from api.views import ProviderListViewSet
+from django.http.response import Http404
+from django.test import RequestFactory
+
+
+class BaseAPITest(APITestCase):
+    def setUp(self):
+        self.factory = RequestFactory()
+        self.view_set = ProviderListViewSet
+        self.base_url = '/api/providers/'
+
+    def get_request_and_response(self, id=None):
+        request = self.factory.get(f'{self.base_url}{id}')
+        if id:
+            return self.view_set.as_view({'get': 'retrieve'})(request, pk=id)
+        return self.view_set.as_view({'get': 'list'})(request)
```

Comment:
you can make this even better by defining a separate get_request method that build the request and two separate methods `retrieve` and `list`. Also you should return the request and response together not only the response.